components/webext-storage/src/sync/outgoing.rs

```diff
@@ -0,0 +1,178 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// The "outgoing" part of syncing - building the payloads to upload and
+// managing the sync state of the local DB.
+
+use interrupt::Interruptee;
+use rusqlite::{Connection, Row, Transaction};
+use sql_support::ConnExt;
+use sync15_traits::ServerTimestamp;
+use sync_guid::Guid as SyncGuid;
+
+use crate::error::*;
+
+use super::ServerPayload;
+
+// This is the "state" for outgoing items - it's meta-data for an item that we
+// use to update the local DB after we POST outgoing records.
+#[derive(Debug)]
+pub struct OutgoingStateHolder {
+    ext_id: String,
+    change_counter: i32,
+}
+
+// A simple holder for the payload that needs to be sent to the servers, and
+// the above item meta-data.
+#[derive(Debug)]
+pub struct OutgoingInfo {
+    pub state: OutgoingStateHolder,
+    pub payload: ServerPayload,
+}
+
+impl OutgoingInfo {
+    fn from_row(row: &Row<'_>) -> Result<Self> {
+        let guid = row
+            .get::<_, Option<SyncGuid>>("guid")?
+            .unwrap_or_else(SyncGuid::random);
+        let ext_id: String = row.get("ext_id")?;
+        let raw_data: Option<String> = row.get("data")?;
+        let (data, deleted) = if raw_data.is_some() {
+            (raw_data, false)
+        } else {
+            (None, true)
+        };
+        Ok(OutgoingInfo {
+            state: OutgoingStateHolder {
```

Comment:
Hmmm, this does mean we'll need the Sync JS side to hold on to this, match them up before calling `setUploaded`, and pass them back. That should work also, but how do you feel about storing all these outgoing ones in a temp table, like Places does it? `guid` and `change_counter` would become columns in that table, and when you mark it as uploaded, doing the write back to the mirror?